src/lib/optimizer/strategy/logical_reduction_rule.cpp

```diff
@@ -0,0 +1,142 @@
+   *      second pass, because manipulating the LQP within `visit_lqp()`, while theoretically possible, is prone to
+   *      bugs.
+   */
+  for (const auto& [predicate_node, flat_conjunction] : predicate_nodes_to_flat_conjunctions) {
+    for (const auto& predicate_expression : flat_conjunction) {
+      lqp_insert_node(predicate_node, LQPInputSide::Left, PredicateNode::make(predicate_expression));
+    }
+    lqp_remove_node(predicate_node);
+  }
+
+  return false;
+}
+
+std::shared_ptr<AbstractExpression> LogicalReductionRule::reduce_distributivity(
+    const std::shared_ptr<AbstractExpression>& input_expression) {
+  // Step 1: `(a AND b AND c) OR (a AND d AND b AND e)` --> `[(a AND b AND c), (a AND d AND b AND e)]`
+  const auto flat_disjunction = flatten_logical_expressions(input_expression, LogicalOperator::Or);
+
+  // Step 2: `[(a AND b AND c), (a AND d AND b AND e)]` --> `[[a, b, c], [a, d, b, e]]`
+  auto flat_disjunction_and_conjunction = std::vector<std::vector<std::shared_ptr<AbstractExpression>>>{};
+  flat_disjunction_and_conjunction.reserve(flat_disjunction.size());
+
+  for (const auto& expression : flat_disjunction) {
+    flat_disjunction_and_conjunction.emplace_back(flatten_logical_expressions(expression, LogicalOperator::And));
+  }
+
+  // Step 3: Identify common_conjunctions: [a, b]
+  auto common_conjunctions = flat_disjunction_and_conjunction.front();
+
+  for (auto conjunction_idx = size_t{1}; conjunction_idx < flat_disjunction_and_conjunction.size(); ++conjunction_idx) {
+    const auto& flat_conjunction = flat_disjunction_and_conjunction[conjunction_idx];
+
+    for (auto common_iter = common_conjunctions.begin(); common_iter != common_conjunctions.end();) {
+      if (std::find_if(flat_conjunction.begin(), flat_conjunction.end(), [&](const auto& expression) {
+            return *expression == *(*common_iter);
+          }) == flat_conjunction.end()) {
+        common_iter = common_conjunctions.erase(common_iter);
+      } else {
+        ++common_iter;
+      }
+    }
+  }
+
+  // Step 4: Remove common_conjunctions from flat_disjunction_and_conjunction.
+  //         flat_disjunction_and_conjunction = [[c], [d, e]]
+  for (auto& flat_conjunction : flat_disjunction_and_conjunction) {
+    for (auto expression_iter = flat_conjunction.begin(); expression_iter != flat_conjunction.end();) {
+      if (std::find_if(common_conjunctions.begin(), common_conjunctions.end(), [&](const auto& expression) {
+            return *expression == *(*expression_iter);
+          }) != common_conjunctions.end()) {
+        expression_iter = flat_conjunction.erase(expression_iter);
+      } else {
+        ++expression_iter;
+      }
+    }
+  }
+
+  // Step 5: Rebuild inflated expression from common_conjunction: `[a, c]` --> `(a AND c)`
+  auto common_conjunction_expression = inflate_logical_expressions(common_conjunctions, LogicalOperator::And);
```

Comment:
Personally, I would do this later, just before `common_conjunction_expression` is actually used